raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
-            // if the response has an error
-            updateVoterConnections(findQuorumResponse);
-        } else {
-            throw new IllegalStateException("Received unexpected response " + response);
+        } else if (error == Errors.INCONSISTENT_GROUP_PROTOCOL) {
+            // For now we treat this as a fatal error. Once we have support for quorum
+            // reassignment, this error could suggest that either we or the recipient of
+            // the request just has stale voter information, which means we can retry
+            // after backing off.
+            throw new IllegalStateException("Received error indicating inconsistent voter sets");
+        } else if (error == Errors.INVALID_REQUEST) {
+            throw new IllegalStateException("Received unexpected invalid request error");
         }
 
-        connection.onResponse(response.correlationId, responseError, currentTimeMs);
-        if (handleNonMatchingResponseLeaderAndEpoch(responseEpoch, responseLeaderId)) {
-            return true;
-        } else if (responseError != Errors.NONE) {
-            if (quorum.isObserver() && responseError == Errors.BROKER_NOT_AVAILABLE)
-                becomeUnattachedFollower(quorum.epoch());
-
-            logger.debug("Discarding response {} with error {}", data, responseError);
-            return true;
-        } else {
-            return false;
-        }
+        return Optional.empty();
     }
 
-    private boolean handleNonMatchingResponseLeaderAndEpoch(int epoch, OptionalInt leaderId) throws IOException {
-        if (epoch < quorum.epoch()) {
-            return true;
+    private void maybeBecomeFollower(
+        OptionalInt leaderId,
+        int epoch
+    ) throws IOException {
+        if (!hasConsistentLeader(epoch, leaderId)) {
+            throw new IllegalStateException("Received response with leader " + leaderId +
+                " and epoch " + epoch + " which is inconsistent with current leader " +
+                quorum.leaderId() + " and epoch " + quorum.epoch());
         } else if (epoch > quorum.epoch()) {
             // For any request type, if the response indicates a higher epoch, we bump our local
             // epoch and become a follower. Responses only include elected leaders.
             GracefulShutdown gracefulShutdown = shutdown.get();
             if (gracefulShutdown != null) {
                 gracefulShutdown.onEpochUpdate(epoch);
             } else {
-                becomeFollower(epoch, leaderId);
+                becomeFollower(leaderId, epoch);
             }
-
-            return true;
-        } else if (hasInconsistentLeader(epoch, leaderId)) {
-            throw new IllegalStateException("Received response with leader " + leaderId +
-                    " which is inconsistent with current leader " + quorum.leaderId());
+        } else if (leaderId.isPresent() && !quorum.leaderId().isPresent()) {
```

Comment:
Since we now have 3 sub-states for a follower:
```
1. voted
2. fetching
3. unattached
```
Not proposing adding another state enum, should we just define 3 separate `is***Follower` check on `QuorumState` which logic is easier to be reused?

Reply:
I added a `hasLeader` method to make this a little more concise.